cobalt/common/libc/tz/parse_posix_tz.cc

```diff
@@ -0,0 +1,463 @@
+// consumed from the input string. This is crucial for the iterative parser
+// to know where to resume parsing.
+template <typename T>
+struct ParseResult {
+  T value;
+  size_t consumed;
+};
+
+std::optional<int> StringViewToInt(std::string_view string_to_convert) {
+  int integer_value;
+  auto [end_pointer, error_code] = std::from_chars(
+      string_to_convert.data(),
+      string_to_convert.data() + string_to_convert.size(), integer_value);
+
+  if (error_code == std::errc()) {
+    return integer_value;
+  }
+  return std::nullopt;
+}
+
+// Parses a sequence of digits from the start of a string_view.
+// On success, returns a ParseResult with the integer value and number of
+// characters consumed. On failure (no digits or invalid number), returns
+// nullopt. This function does not modify the input string_view.
+std::optional<ParseResult<int>> ParseInteger(std::string_view sv) {
+  size_t digits_end = 0;
+  while (digits_end < sv.length() && isdigit(sv[digits_end])) {
+    digits_end++;
+  }
+
+  if (digits_end == 0) {
+    return std::nullopt;
+  }
+
+  if (auto parsed_value = StringViewToInt(sv.substr(0, digits_end))) {
+    return {{*parsed_value, digits_end}};
+  }
+
+  return std::nullopt;
+}
+
+// Parses an optional, colon-prefixed time component (minutes or seconds).
+// Returns `true` on success (if the component was parsed, or if it wasn't
+// present). Returns `false` on a parse error (e.g., ":" not followed by
+// digits). On success, it updates `sv`, `consumed`, and `target_component`.
+bool ParseOptionalTimeComponent(std::string_view& sv,
+                                size_t& consumed,
+                                int& target_component) {
+  if (sv.empty() || sv.front() != ':') {
+    return true;  // Success: component is optional and not present.
+  }
+
+  // We have a ':', so a number must follow.
+  if (auto int_res = ParseInteger(sv.substr(1))) {
+    target_component = int_res->value;
+    // Consume the ':' plus the digits.
+    size_t component_consumed = 1 + int_res->consumed;
+    sv.remove_prefix(component_consumed);
+    consumed += component_consumed;
+    return true;
+  }
+
+  // Parse error: ':' was not followed by valid digits.
+  return false;
+}
+
+// Parses a time offset string (e.g., "8", "+2", or "-1:30") into seconds.
+// Returns the offset in seconds and the number of characters consumed.
+std::optional<ParseResult<TimeOffset>> ParseTimeOffset(
+    std::string_view time_offset_string) {
+  if (time_offset_string.empty()) {
+    return std::nullopt;
+  }
+
+  bool is_negative = false;
+  size_t consumed = 0;
+
+  // Handle optional leading '+' or '-' sign.
+  if (time_offset_string.front() == '+') {
+    time_offset_string.remove_prefix(1);
+    consumed++;
+  } else if (time_offset_string.front() == '-') {
+    is_negative = true;
+    time_offset_string.remove_prefix(1);
+    consumed++;
+  }
+
+  // --- Parse Hours (mandatory) ---
+  int total_seconds = 0;
```

Comment:
nit: preference for early return and non-nested code otherwise

```
  auto hours_res = ParseInteger(time_offset_string);
  if (!hours_res) {
    return std::nullopt;
  }
  int total_seconds += hours_res->value * kHour;
...
```